agent/src/com/cloud/agent/direct/download/HttpDirectTemplateDownloader.java

```diff
@@ -0,0 +1,158 @@
+        super(url, destPoolPath, templateId, checksum);
+        client = new HttpClient(s_httpClientManager);
+        myretryhandler = createRetryTwiceHandler();
+        request = createRequest(url, headers);
+        String downloadDir = getDirectDownloadTempPath(templateId);
+        createTemporaryDirectoryAndFile(downloadDir);
+    }
+
+    private void createTemporaryDirectoryAndFile(String downloadDir) {
+        createFolder(getDestPoolPath() + File.separator + downloadDir);
+        File f = new File(getDestPoolPath() + File.separator + downloadDir + File.separator + getFileNameFromUrl());
+        setDownloadedFilePath(f.getAbsolutePath());
+    }
+
+    private GetMethod createRequest(String downloadUrl, Map<String, String> headers) {
+        GetMethod request = new GetMethod(downloadUrl);
+        request.getParams().setParameter(HttpMethodParams.RETRY_HANDLER, myretryhandler);
+        request.setFollowRedirects(true);
+        for (String key : headers.keySet()) {
+            request.setRequestHeader(key, headers.get(key));
+        }
+        return request;
+    }
+
+    private long getRemoteSize(GetMethod request) {
+        Header contentLengthHeader = request.getResponseHeader("Content-Length");
+        long reportedRemoteSize = 0;
+        if (contentLengthHeader == null) {
+            Header chunkedHeader = request.getResponseHeader("Transfer-Encoding");
+            if (chunkedHeader == null || !"chunked".equalsIgnoreCase(chunkedHeader.getValue())) {
+                throw new CloudRuntimeException("Error template remote size");
+            }
+        } else {
+            reportedRemoteSize = Long.parseLong(contentLengthHeader.getValue());
+        }
+        return reportedRemoteSize;
+    }
+
+    private HttpMethodRetryHandler createRetryTwiceHandler() {
+        return new HttpMethodRetryHandler() {
+            @Override
+            public boolean retryMethod(final HttpMethod method, final IOException exception, int executionCount) {
+                if (executionCount >= 2) {
+                    // Do not retry if over max retry count
+                    return false;
+                }
+                if (exception instanceof NoHttpResponseException) {
+                    // Retry if the server dropped connection on us
+                    return true;
+                }
+                if (!method.isRequestSent()) {
+                    // Retry if the request has not been sent fully or
+                    // if it's OK to retry methods that have been sent
+                    return true;
+                }
+                // otherwise do not retry
+                return false;
+            }
+        };
+    }
+
+    @Override
+    public boolean downloadTemplate() {
+        InputStream in = null;
+        RandomAccessFile out = null;
+        try {
+            client.executeMethod(request);
+            remoteSize = getRemoteSize(request);
```

Comment:
`remoteSize` is never read

Reply:
Hi @marcaurele, in fact it is, on copyBytes method in line 141

Reply:
Since you're using this info in the `copyBytes` method, this is overkill. It can even throw an exception but having the filesize isn't mandatory to download it. This information, if present in the header, should be used as a *nice to have* but the download **must work** without this info as @wido pointed that some servers don't send this information in the header (https://github.com/apache/cloudstack/pull/1934#issuecomment-347784289). So it must be changed